website/sales/models/order.py

```diff
@@ -0,0 +1,271 @@
+                        Exists(
+                            OrderItem.objects.filter(
+                                order=OuterRef("pk"),
+                                product__product__age_restricted=True,
+                            )
+                        ),
+                        then=Value(True),
+                    ),
+                    default=Value(False),
+                    output_field=BooleanField(),
+                ),
+            )
+        )
+
+
+class Order(models.Model, Payable):
+    objects = OrderManager()
+
+    class Meta:
+        verbose_name = "order"
+        verbose_name_plural = "orders"
+        permissions = [
+            ("custom_prices", "Can use custom prices and discounts in orders"),
+        ]
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+
+    created_at = models.DateTimeField(default=timezone.now)
+
+    # created_by = models.ForeignKey(Member, on_delete=SET_NULL, related_name='orders_created',) # TODO add this field
+
+    shift = models.ForeignKey(
+        Shift,
+        related_name="orders",
+        default=default_order_shift,
+        null=False,
+        blank=False,
+        on_delete=models.PROTECT,
+    )
+
+    items = models.ManyToManyField(ProductListItem, through="OrderItem")
+
+    payment = models.OneToOneField(
+        Payment,
+        related_name="sales_order",
+        on_delete=models.SET_NULL,
+        blank=True,
+        null=True,
+    )
+
+    discount = models.DecimalField(
+        max_digits=6,
+        decimal_places=2,
+        null=True,
+        blank=True,
+        validators=[MinValueValidator(Decimal("0.00"))],
+    )
+
+    payer = models.ForeignKey(
+        Member, models.SET_NULL, related_name="sales_order", blank=True, null=True,
+    )
+
+    def save(
+        self, force_insert=False, force_update=False, using=None, update_fields=None
+    ):
+        if self.shift.locked:
+            return
+        if self.payment and self.get_total_amount() != self.payment.amount:
+            return
+        if self.payment and not self.payer:
+            self.payer = self.payment.paid_by
+
+        return super(Order, self).save(force_insert, force_update, using, update_fields)
+
+    def clean(self):
+        super().clean()
+        errors = {}
+
+        if self.shift.locked:
+            errors.update({"shift": _("The shift this order belongs to is locked.")})
+
+        if self.discount and self.discount > self.get_total_amount():
+            errors.update(
+                {"discount": _("Discount cannot be higher than total amount.")}
+            )
+
+        if errors:
+            raise ValidationError(errors)
+
+    def get_age_restricted(self):
+        return any(
+            self.order_items.values_list("product__product__age_restricted", flat=True)
+        )
+
+    def get_subtotal(self):
+        qs = self.order_items.aggregate(subtotal=Coalesce(Sum("total"), Value(0.00)))
```

Comment:
I was wondering, since you replace the default queryset manager by the one with the annotations do you not already have this information for every time you get the models? Since it is essentially already part of the queryset.

Reply:
Yes and no. When you materialize an Order from a QuerySet, the shitty thing is that the annotations get lost. It is really ugly. For example, in the API I need to turn all those annotations into serializermethodfields, because when querying an order the annotations are present, but on updating them, the order is materialized to a python object and suddenly the annotated fields don't exist anymore. That's exactly what django-queryable-properties tries to fix (and does fix quite decently actually): it allows you to define python properties in terms of Django ORM calculus and that works fine (but the problem with that package is that you can't use the annotations without explicitly using `.select_properties(()` on the QuerySet).

The reason why I still can use the `order_items.aggregate` here is that `order_items` will return a new manager whatsoever.